client/src/components/dialog/Dialog.tsx

```diff
@@ -0,0 +1,73 @@
+import { ReactNode, useEffect, useRef, useState } from "react";
+
+import "./Dialog.css";
+
+
+interface UseDialogResult {
+    isOpen: boolean;
+    close: () => void;
+    open: () => void;
+}
+
+export function useDialog(): UseDialogResult {
+    const [isOpen, setIsOpen] = useState<boolean>(false);
+
+    const open = () => {
+        setIsOpen(true);
+    }
+
+    const close = () => {
+        setIsOpen(false);
+    };
+
+    return { isOpen, open, close };
+}
+
+
+interface DialogProps {
+    isOpen: boolean;
+    onClose: () => void;
+    children: ReactNode;
+}
+
+function Dialog({ isOpen, onClose, children }: DialogProps) {
+
+    const dialogRef = useRef<HTMLDialogElement>(null);
+
+    const handleOutsideClick = (event: MouseEvent) => {
```

Comment:
Do you need to wrap this in `useCallback`? Whenever the component renders, I think the `handleOutsideClick` function would be different since it is a lambda function. Thus, when you call `document.removeEventListener`, I am not sure that the initial `handleOutsideClick` passed in during `document.addEventListener` would actually be removed.

Reply:
Nice catch. Thanks for flagging.